modules/shield/handle_msg.go

```diff
@@ -0,0 +1,188 @@
+
+	case *shieldtypes.MsgUnstakeFromShield:
+		return m.HandleMsgUnstakeFromShield(tx, cosmosMsg)
+
+	}
+
+	return nil
+}
+
+// HandleMsgCreatePool allows to properly handle a MsgCreatePool
+func (m *Module) HandleMsgCreatePool(index int, tx *juno.Tx, msg *shieldtypes.MsgCreatePool) error {
+
+	// Get create pool event
+	createPoolEvent, err := tx.FindEventByType(index, "create_pool")
+	if err != nil {
+		return fmt.Errorf("error while getting create pool event: %s", err)
+	}
+
+	// Get pool ID
+	poolIDStr, err := tx.FindAttributeByKey(createPoolEvent, shieldtypes.AttributeKeyPoolID)
+	if err != nil {
+		return fmt.Errorf("error while getting pool ID from MsgCreatePool: %s", err)
+	}
+
+	// Convert pool ID
+	poolID, err := strconv.ParseUint(poolIDStr, 10, 64)
+	if err != nil {
+		return fmt.Errorf("error while parsing pool ID to uint64: %s", err)
+	}
+
+	pool := types.NewShieldPool(
+		poolID, msg.Shield[0].Amount, msg.Deposit.Native, msg.Deposit.Foreign, msg.Sponsor,
+		msg.SponsorAddr, msg.Description, msg.ShieldLimit, false, tx.Height,
+	)
+
+	return m.db.SaveShieldPool(pool)
+}
+
+// HandleMsgUpdatePool allows to properly handle a MsgUpdatePool
+func (m *Module) HandleMsgUpdatePool(tx *juno.Tx, msg *shieldtypes.MsgUpdatePool) error {
+
+	// Sponsor, sponsor address, and pool pause status will not be updated with ON CONFLICT statement
+	pool := types.NewShieldPool(
+		msg.PoolId, msg.Shield[0].Amount, msg.ServiceFees.Native, msg.ServiceFees.Foreign, "",
+		"", msg.Description, msg.ShieldLimit, false, tx.Height,
+	)
+
+	return m.db.SaveShieldPool(pool)
+}
+
+// HandleMsgPausePool allows to properly handle a MsgPausePool
+func (m *Module) HandleMsgPausePool(tx *juno.Tx, msg *shieldtypes.MsgPausePool) error {
+	pause := true
+	return m.db.UpdatePoolPauseStatus(msg.PoolId, pause, tx.Height)
+}
+
+// HandleMsgResumePool allows to properly handle a MsgResumePool
+func (m *Module) HandleMsgResumePool(tx *juno.Tx, msg *shieldtypes.MsgResumePool) error {
+	pause := false
+	return m.db.UpdatePoolPauseStatus(msg.PoolId, pause, tx.Height)
+}
+
+// HandleMsgWithdrawRewards allows to properly handle a MsgWithdrawRewards
+func (m *Module) HandleMsgWithdrawRewards(tx *juno.Tx, msg *shieldtypes.MsgWithdrawRewards) error {
+	return m.db.WithdrawNativeRewards(msg.From, tx.Height)
+}
+
+// HandleMsgWithdrawForeignRewards allows to properly handle a MsgWithdrawForeignRewards
+func (m *Module) HandleMsgWithdrawForeignRewards(tx *juno.Tx, msg *shieldtypes.MsgWithdrawForeignRewards) error {
+	return m.db.WithdrawForeignRewards(msg.From, tx.Height)
+}
+
+// HandleMsgDepositCollateral allows to properly handle a MsgDepositCollateral
+func (m *Module) HandleMsgDepositCollateral(tx *juno.Tx, msg *shieldtypes.MsgDepositCollateral) error {
+	collateral, err := m.db.GetShieldProviderCollateral(msg.From)
+	if err != nil {
+		return fmt.Errorf("error while getting shield provider collateral: %s", err)
+	}
+	updatedCollateral := collateral + msg.Collateral[0].Amount.Int64()
```

Comment:
would it be better if we use `shieldClient.Provider()` to query the total collateral?